In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
dt = pd.read_csv('/content/titanic_dataset .csv')

In [3]:
dt

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [5]:
dt.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [6]:
dt= dt.drop('Cabin',axis=1)

In [9]:
dt['Age'].fillna(dt['Age'].mean(),inplace=True)

In [10]:
dt['Embarked'].fillna(dt['Embarked'].mode()[0],inplace=True)

In [11]:
dt.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [32]:
y=dt['Survived']

In [33]:
x=dt.drop('Survived',axis=1)

In [26]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
dt['Embarked'] = label_encoder.fit_transform(dt['Embarked'])
dt['Name'] = label_encoder.fit_transform(dt['Name'])
dt['Sex'] = label_encoder.fit_transform(dt['Sex'])
dt['Ticket'] = label_encoder.fit_transform(dt['Ticket'])


In [31]:
dt.head()

PassengerId  Survived  Pclass  Name  Sex   Age  SibSp  Parch  Ticket  \
0            1         0       3   108    1  22.0      1      0     523   
1            2         1       1   190    0  38.0      1      0     596   
2            3         1       3   353    0  26.0      0      0     669   
3            4         1       1   272    0  35.0      1      0      49   
4            5         0       3    15    1  35.0      0      0     472   

      Fare  Embarked  
0   7.2500         2  
1  71.2833         0  
2   7.9250         2  
3  53.1000         2  
4   8.0500         2

In [28]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [34]:
x = scaler.fit_transform(x)

In [36]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state = 42, test_size = 0.2)

KNN

In [37]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)

KNeighborsClassifier()

In [38]:
score_knn=knn.score(x_test,y_test)

In [39]:
score_knn

0.8100558659217877

SVM

In [40]:
from sklearn.svm import SVC
svc= SVC()
svc.fit(x_train, y_train)

SVC()

In [41]:
score_svc=svc.score(x_test,y_test)

In [43]:
score_svc

0.8100558659217877

Kfold

In [45]:
from sklearn.model_selection import KFold
kfold_validator  = KFold(10)
for train_index,test_index in kfold_validator.split(x,y):
    print('Training Index: ', train_index)
    print('Testing Index: ',test_index)

Training Index:  [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251
 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269
 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287
 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 305
 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321 322 323
 324 325 326 327 328 329 330 331 3

In [47]:
from sklearn.model_selection import cross_val_score
cross_val_result = cross_val_score(knn,x,y,cv = kfold_validator)

In [48]:
cross_val_result

array([0.76666667, 0.80898876, 0.78651685, 0.84269663, 0.85393258,
       0.73033708, 0.82022472, 0.80898876, 0.84269663, 0.82022472])

In [49]:
np.mean(cross_val_result)

0.8081273408239701

In [66]:
from sklearn.model_selection import cross_val_score
cross_val_result2 = cross_val_score(svc,x,y,cv = kfold_validator)

In [67]:
cross_val_result2

array([0.8       , 0.86516854, 0.78651685, 0.84269663, 0.83146067,
       0.80898876, 0.79775281, 0.76404494, 0.85393258, 0.83146067])

In [52]:
np.mean(cross_val_result2)

0.8182022471910113

**Stratified cross validation**

In [53]:
from sklearn.model_selection import StratifiedKFold

In [55]:
skfold_validator = StratifiedKFold(n_splits = 10)

In [56]:
for train_index,test_index in skfold_validator.split(x,y):
    print('Training Index: ', train_index)
    print('Testing Index: ',test_index)

Training Index:  [ 82  84  85  88  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251
 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269
 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287
 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 305
 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321 322 323
 324 325 326 327 328 329 330 331 3

In [58]:
scross_val_result1 = cross_val_score(knn,x,y,cv = skfold_validator)

In [59]:
scross_val_result1

array([0.76666667, 0.7752809 , 0.7752809 , 0.82022472, 0.86516854,
       0.73033708, 0.84269663, 0.80898876, 0.80898876, 0.82022472])

In [60]:
np.mean(scross_val_result1)

0.801385767790262

In [69]:
scross_val_result2 =cross_val_score(svc,x,y,cv = skfold_validator)

In [70]:
scross_val_result2

array([0.81111111, 0.85393258, 0.76404494, 0.87640449, 0.84269663,
       0.79775281, 0.82022472, 0.7752809 , 0.84269663, 0.85393258])

In [71]:
np.mean(scross_val_result2)

0.8238077403245943

**Average acuuracy score of KFOLD**

KNN = 0.8081273408239701






SVM=0.8182022471910113


**Average accuracy score of Stratified**

KNN =0.801385767790262


SVM = 0.8238077403245943